File: lua/puzzle2-2.lua
require "utils"

local sum = 0
for line in io.lines() do
	local game, sets = unpack(string.split(line, ":"))
	assert(game and sets)

	sets = string.strip(sets)
	max = {red=1, green=1, blue=1}
	for _, cubes in ipairs(string.split(sets, "[,;] ")) do
		local n, color = string.match(cubes, "(%d+) (%a+)")
		assert(n and color)
		n = n + 0
		if n > max[color] then max[color] = n end
	end

	sum = sum + max.red * max.green * max.blue
end
print(sum)

File: rust/src/bin/puzzle2-2.rs
fn main() {
    let mut sum: u32 = 0;
    for line in std::io::stdin().lines().map(|x| x.unwrap()) {
        let mut max_cubes = std::collections::HashMap::new();
        let (_, sets) = line.split_once(": ").unwrap();
        for set in sets.split("; ") {
            for cubes in set.split(", ") {
                let (n, color) = cubes.split_once(' ').unwrap();
                let n: u32 = n.parse().unwrap();
                let max = max_cubes.entry(color).or_insert(n);
                if n > *max {
                    *max = n;
                }
            }
        }
        sum += max_cubes.values().product::<u32>();
    }
    println!("{sum}");
}
